File: YuXiangROS/Chap7/Action_ws/src/demo_python_action/demo_python_action/action_client.py
import random

import rclpy


# 用于判断目标执行状态
from action_msgs.msg import GoalStatus

# 导入自定义的动作接口类型
from chap7_interfaces.action import NavigateToPose
from rclpy.action import ActionClient
from rclpy.node import Node


class NavActionClient(Node):
    def __init__(self):
        super().__init__("nav_action_client")

        # 创建动作客户端: 动作名 /navigate_to_pose, 类型 NavigateToPose
        self.action_client_ = ActionClient(self, NavigateToPose, "/navigate_to_pose")

        # 当前目标句柄与状态标记
        self.goal_handle_ = None
        self.feedback_count_ = 0
        # 反馈达到该次数后取消目标, 演示动作的可取消性
        self.feedback_threshold_ = 5
        self.cancel_sent_ = False
        self.goal_done_ = True

        # 创建定时器: 每 10 秒发送一个新目标 (异步, 不阻塞主线程)
        self.send_goal_timer_ = self.create_timer(10.0, self.send_new_goal)

    def send_new_goal(self):
        """
        定时器回调函数, 发送新目标
        """

        # 上一目标未结束, 跳过本轮
        if not self.goal_done_:
            return
        # 等待动作服务端启动
        if not self.action_client_.wait_for_server(timeout_sec=1.0):
            self.get_logger().info("Server not ready, waiting...")
            return

        # 随机生成目标点
        goal = NavigateToPose.Goal()
        goal.target_x = random.uniform(1.0, 10.0)
        goal.target_y = random.uniform(1.0, 10.0)

        # 重置状态
        self.goal_done_ = False
        self.cancel_sent_ = False
        self.feedback_count_ = 0

        self.get_logger().info(
            f"Sending goal: ({goal.target_x:.2f}, {goal.target_y:.2f}), "
            f"cancel after {self.feedback_threshold_} feedbacks"
        )

        # 注册反馈回调后异步发送目标
        self.send_goal_future_ = self.action_client_.send_goal_async(goal, feedback_callback=self.feedback_callback)
        self.send_goal_future_.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future):
        """
        目标响应回调函数
        Args:
            future: 异步结果
        """

        goal_handle = future.result()
        # 目标被拒绝时返回空句柄
        if not goal_handle.accepted:
            self.get_logger().info("Goal rejected by server")
            self.goal_done_ = True
            return

        # 记录目标句柄, 用于后续取消
        self.goal_handle_ = goal_handle
        self.get_logger().info("Goal accepted by server")
        # 注册结果回调
        self.result_future_ = goal_handle.get_result_async()
        self.result_future_.add_done_callback(self.result_callback)

    def feedback_callback(self, feedback_msg):
        """
        反馈回调函数: 打印进度, 累计一定次数后取消目标
        Args:
            feedback_msg: 反馈消息, 类型 NavigateToPose.Feedback
        """

        feedback = feedback_msg.feedback
        self.get_logger().info(
            f"Feedback: pos ({feedback.current_x:.2f}, {feedback.current_y:.2f}), "
            f"dist {feedback.distance:.2f}, progress {feedback.progress * 100:.2f}%"
        )

        # 累计反馈次数, 达到阈值后发送取消请求
        self.feedback_count_ += 1
        if self.feedback_count_ == self.feedback_threshold_ and not self.cancel_sent_:
            self.cancel_sent_ = True
            self.get_logger().info(f"{self.feedback_threshold_} feedbacks, sending cancel...")
            # 发送取消请求 (对应 C++ 版 action_client_->async_cancel_goal)
            self.goal_handle_.cancel_goal_async()

    def result_callback(self, future):
        """
        结果回调函数
        Args:
            future: 异步结果
        """

        # 标记当前目标结束
        self.goal_done_ = True

        # 根据状态码打印执行结果
        status = future.result().status
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Goal succeeded")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().info("Goal aborted")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().info("Goal canceled")
        else:
            self.get_logger().info(f"Unknown goal status: {status}")

        # 打印自定义结果字段
        action_result = future.result().result
        self.get_logger().info(
            f"Result: result={action_result.result}, final_distance: {action_result.final_distance:.2f}"
        )


def main(args=None):
    rclpy.init(args=args)

    node = NavActionClient()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: YuXiangROS/Chap7/Action_ws/src/demo_cpp_action/src/action_client.cpp
#include <chap7_interfaces/action/navigate_to_pose.hpp> // 导入自定义的动作接口类型
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp> // 动作通信库

#include <memory>
#include <random>

using namespace std::chrono_literals;

using NavigateToPose = chap7_interfaces::action::NavigateToPose;
using GoalHandleNavigate = rclcpp_action::ClientGoalHandle<NavigateToPose>;

class NavActionClient : public rclcpp::Node {
  private:
    // 动作客户端智能指针: 动作名 /navigate_to_pose, 类型 NavigateToPose
    rclcpp_action::Client<NavigateToPose>::SharedPtr action_client_;
    // 定时器: 周期性发送目标, 演示动作的异步性
    rclcpp::TimerBase::SharedPtr send_goal_timer_;
    // 当前目标句柄
    GoalHandleNavigate::SharedPtr goal_handle_;
    // 随机数生成器, 用于生成目标点
    std::mt19937 generator_;
    // 目标点坐标分布: [1.0, 10.0]
    std::uniform_real_distribution<double> coordinate_distribution_{1.0, 10.0};
    // 反馈累计次数
    int feedback_count_{0};
    // 反馈达到该次数后取消目标, 演示动作的可取消性
    int feedback_threshold_{5};
    // 是否已发送取消请求
    bool cancel_sent_{false};
    // 当前目标是否已结束
    bool goal_done_{true};

  public:
    NavActionClient() : Node("nav_action_client"), generator_(std::random_device{}()) {
        // 创建动作客户端: 动作名 /navigate_to_pose, 类型 NavigateToPose
        action_client_ = rclcpp_action::create_client<NavigateToPose>(this, "/navigate_to_pose");
        // 创建定时器: 每 10 秒发送一个新目标 (异步, 不阻塞主线程)
        send_goal_timer_ =
            this->create_wall_timer(10s, std::bind(&NavActionClient::send_new_goal_, this));
    }

  private:
    /**
     * @brief 定时器回调函数: 发送新目标
     */
    void send_new_goal_() {
        // 上一目标未结束, 跳过本轮
        if (goal_done_ == false) {
            return;
        }
        // 等待动作服务端启动
        if (this->action_client_->wait_for_action_server(1s) == false) {
            RCLCPP_INFO(this->get_logger(), "Server not ready, waiting...");
            return;
        }

        // 随机生成目标点
        auto goal = NavigateToPose::Goal();
        goal.target_x = coordinate_distribution_(generator_);
        goal.target_y = coordinate_distribution_(generator_);

        // 重置状态
        goal_done_ = false;
        cancel_sent_ = false;
        feedback_count_ = 0;

        RCLCPP_INFO(
            this->get_logger(),
            "Sending goal: (%.2f, %.2f), cancel after %d feedbacks",
            goal.target_x,
            goal.target_y,
            feedback_threshold_
        );

        // 注册三个回调后异步发送目标
        auto send_goal_options = rclcpp_action::Client<NavigateToPose>::SendGoalOptions();
        send_goal_options.goal_response_callback =
            std::bind(&NavActionClient::goal_response_callback_, this, std::placeholders::_1);
        send_goal_options.feedback_callback = std::bind(
            &NavActionClient::feedback_callback_,
            this,
            std::placeholders::_1,
            std::placeholders::_2
        );
        send_goal_options.result_callback =
            std::bind(&NavActionClient::result_callback_, this, std::placeholders::_1);
        this->action_client_->async_send_goal(goal, send_goal_options);
    }

    /**
     * @brief 目标响应回调函数
     * @param goal_handle: 目标句柄, 为空表示被拒绝
     */
    void goal_response_callback_(const GoalHandleNavigate::SharedPtr& goal_handle) {
        if (goal_handle != nullptr) {
            goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(), "Goal accepted by server");
        } else {
            RCLCPP_INFO(this->get_logger(), "Goal rejected by server");
            goal_done_ = true;
        }
    }

    /**
     * @brief 反馈回调函数: 打印进度, 累计一定次数后取消目标
     * @param goal_handle: 目标句柄
     * @param feedback: 反馈消息, 类型 NavigateToPose::Feedback
     */
    void feedback_callback_(
        const GoalHandleNavigate::SharedPtr /* goal_handle */,
        const std::shared_ptr<const NavigateToPose::Feedback> feedback
    ) {
        RCLCPP_INFO(
            this->get_logger(),
            "Feedback: pos (%.2f, %.2f), dist %.2f, progress %.2f%%",
            feedback->current_x,
            feedback->current_y,
            feedback->distance,
            feedback->progress * 100
        );

        // 累计反馈次数, 达到阈值后发送取消请求
        feedback_count_++;
        if (feedback_count_ == feedback_threshold_ && cancel_sent_ == false) {
            cancel_sent_ = true;
            RCLCPP_INFO(this->get_logger(), "%d feedbacks, sending cancel...", feedback_threshold_);
            this->action_client_->async_cancel_goal(goal_handle_);
        }
    }

    /**
     * @brief 结果回调函数
     * @param result: 结果消息, 类型 NavigateToPose::Result
     */
    void result_callback_(const GoalHandleNavigate::WrappedResult& result) {
        goal_done_ = true;
        // 根据状态码打印执行结果
        switch (result.code) {
        case rclcpp_action::ResultCode::SUCCEEDED:
            RCLCPP_INFO(this->get_logger(), "Goal succeeded");
            break;
        case rclcpp_action::ResultCode::ABORTED:
            RCLCPP_INFO(this->get_logger(), "Goal aborted");
            return;
        case rclcpp_action::ResultCode::CANCELED:
            RCLCPP_INFO(this->get_logger(), "Goal canceled");
            break;
        default:
            RCLCPP_INFO(this->get_logger(), "Unknown goal status");
            return;
        }
        // 打印自定义结果字段
        RCLCPP_INFO(
            this->get_logger(),
            "Result: result=%d, final_distance: %.2f",
            result.result->result,
            result.result->final_distance
        );
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NavActionClient>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
